data/kernels/sigmoid.cl

```diff
@@ -0,0 +1,291 @@
+/*
+    This file is part of darktable,
+    copyright (c) 2018-2020 darktable developers.
+
+    darktable is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    darktable is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "common.h"
+
+typedef struct dt_iop_sigmoid_value_order_t
+{
+  size_t min;
+  size_t mid;
+  size_t max;
+} dt_iop_sigmoid_value_order_t;
+
+static void pixel_channel_order(const float *pix_in, dt_iop_sigmoid_value_order_t *pixel_value_order)
+{
+  if (pix_in[0] >= pix_in[1])
+  {
+    if (pix_in[1] > pix_in[2])
+    {  // Case 1: r >= g >  b
+      pixel_value_order->max = 0;
+      pixel_value_order->mid = 1;
+      pixel_value_order->min = 2;
+    }
+    else if (pix_in[2] > pix_in[0])
+    {  // Case 2: b >  r >= g
+      pixel_value_order->max = 2;
+      pixel_value_order->mid = 0;
+      pixel_value_order->min = 1;
+    }
+    else if (pix_in[2] > pix_in[1])
+    {  // Case 3: r >= b >  g
+      pixel_value_order->max = 0;
+      pixel_value_order->mid = 2;
+      pixel_value_order->min = 1;
+    }
+    else
+    {  // Case 4: r == g == b
+      // No change of the middle value, just assign something.
+      pixel_value_order->max = 0;
+      pixel_value_order->mid = 1;
+      pixel_value_order->min = 2;
+    }
+  }
+  else
+  {
+    if (pix_in[0] >= pix_in[2])
+    {  // Case 5: g >  r >= b
+      pixel_value_order->max = 1;
+      pixel_value_order->mid = 0;
+      pixel_value_order->min = 2;
+    }
+    else if (pix_in[2] > pix_in[1])
+    {  // Case 6: b >  g >  r
+      pixel_value_order->max = 2;
+      pixel_value_order->mid = 1;
+      pixel_value_order->min = 0;
+    }
+    else
+    {  // Case 7: g >= b >  r
+      pixel_value_order->max = 1;
+      pixel_value_order->mid = 2;
+      pixel_value_order->min = 0;
+    }
+  }
+}
+
+static inline float4 desaturate_negative_values(const float4 i)
+{
+  const float pixel_average = fmax((i.x + i.y + i.z) / 3.0f, 0.0f);
+  const float min_value = fmin(fmin(i.x, i.y), i.z);
+  const float saturation_factor = min_value < 0.0f ? -pixel_average / (min_value - pixel_average) : 1.0f;
+
+  return pixel_average + saturation_factor * (i - pixel_average);
+}
+
+static inline float generalized_loglogistic_sigmoid_scalar(const float value, const float magnitude, const float paper_exp,
```

Comment:
some style changes to follow dt convention.

- static routines must start with underscore
- the parameters one per line when there is too many to fit on a line

So it should be:

```
static inline float _generalized_loglogistic_sigmoid_scalar(const float value, 
                                                            const float magnitude, 
                                                            const float paper_exp,
                                                            const float film_fog, 
                                                            const float film_power, 
                                                            const float paper_power)
{
```

Many other occurrences in following code.